lib/pbench/server/cache_manager.py

```diff
@@ -82,6 +95,59 @@ def __str__(self) -> str:
         return f"An error occurred while changing file permissions of {self.tarball}: {self.error}"
 
 
+class CacheType(Enum):
+
+    FILE = auto()
+    DIRECTORY = auto()
+    SYMLINK = auto()
+    OTHER = auto()
+
+
+@dataclass
+class CacheObject:
+
+    name: str = None
+    location: Path = None
+    resolve_path: Optional[Path] = None
+    resolve_type: str = None
+    size: int = None
+    type: str = CacheType
+    dir_path: InitVar[Path] = None
+    path: InitVar[Path] = None
+
+    def __post_init__(self, dir_path, path):
+        self.name = path.name
+        self.location = path.relative_to(dir_path)
+        if path.is_symlink():
+            self.type = CacheType.SYMLINK
+            try:
+                resolve_path = path.resolve(strict=True)
+            except FileNotFoundError:
+                resolve_path = path.readlink()
+
+            try:
+                self.resolve_path = resolve_path.relative_to(dir_path)
+            except ValueError:
+                self.resolve_path = resolve_path
+                self.resolve_type = CacheType.SYMLINK
+            else:
+                if not resolve_path.exists():
+                    self.resolve_type = CacheType.SYMLINK
```

Comment:
I removed that because I wasn't able to understand the relevance of it. The `resolve_path` will be either resolved or it will be a readlink in this try block. When we try to find the relative path here and it fails, it might be due to many reasons. if the path is outside the scope if this directory like `/etc/passwd` we do not want to add `dir_path` infront of it. we just want the readlink to be added there and not do something else with it. I dont't think adding `dir_path` infront of an unresolved path is making anything easier.

Reply:
As I mentioned elsewhere, there is a matrix of symlink cases which we should cover in the testing:
- the reference is either absolute or relative,
- it points either within the results area or outside it,
- it refers to a directory, a regular file, or something else, and
- the referent either exists or it doesn't.

Fortunately, I think we can collapse a bunch of these cases:
- If the original reference is absolute, then, while it might have made sense on the host where the results were collected, we can't guarantee that it will make sense in the cache, so we should mark that resolution as `OTHER`.
- If the reference points outside the results area, then it refers to a file on the host where the results were collected, so we cannot know what sort of file it was, so we should mark it as `OTHER`.
- If the reference doesn't exist on our local machine, then we can't know what it's type is/was (if it ever existed at all), so mark it as `OTHER`.

So, that leaves us with:
- the reference must be a relative path,
- it must resolve to a file (or directory) within the results area (and it must do so without so many `..` references that it leaves the results area and manages to return to it -- that would require the same "coincidence" which is the basis for us excluding absolute paths)
- and it must exist in the results area on the local machine.

So, we read the link file and confirm that it contains a relative path; then we use `Path.resolve()` on it to locate the file to which it (perhaps via other symlinks) refers (and to prove that that file exists); then we use `Path.relative_to()` to generate the location in the results area relative to the cache root.

Assuming that none of those operations result in an error, now we have two paths to the referent:  we have the absolute path produced by `Path.resolve()` and we have the relative path produced by `Path.relative_to()`.  We can use the first one to determine what kind of file the referent is.

The problem with the code above is that, if the `Path.resolve()` operation succeeds, the `resolve_path` will be an absolute path, but, if it fails, then we will set the `resolve_path` to the contents of the symlink file which might be a relative path!  If `resolve_path` is a relative path, then, when we call `Path.exists()`, the relative path evaluation will be performed relative to the current working directory...so, we're sure to get the wrong result.  The way to address this is to make sure that we're working with a relative path in all cases, to ensure that it is always relative to the same directory, and to prefix it with that directory's path before doing the file checks...which is what the code that you left out was supposed to do (although, thinking about it now, I don't think it quite accomplished it 🥴).

However, in [the code that I'm currently proposing we use](https://github.com/distributed-system-analysis/pbench/pull/3396#discussion_r1198353365), we no longer directly use the contents of the symlink file for anything other than to report when we cannot resolve the link to another file, so the concern about the missing code-bit will go away if you adopt my suggestion.  🙂

Reply:
> in the code that I'm currently proposing we use, we no longer directly use the contents of the symlink file for anything other than to report when we cannot resolve the link to another file, so the concern about the missing code-bit will go away if you adopt my suggestion.

Actually, you've already accepted the part of the suggestion which addresses this problem...so we're good on that point.  🙂